File: app/src/main/java/me/rohank05/beat/lavaplayer/TrackScheduler.java
package me.rohank05.beat.lavaplayer;

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.TextChannel;

import java.util.ArrayList;
import java.util.concurrent.TimeUnit;

public class TrackScheduler extends AudioEventAdapter {
    public final AudioPlayer player;
    public final ArrayList<AudioTrack> queue;
    public final TextChannel channel;

    public TrackScheduler(AudioPlayer player, TextChannel channel) {
        this.player = player;
        this.queue = new ArrayList<>();
        this.channel = channel;
    }

    public void queue(AudioTrack track) {
        if (!this.player.startTrack(track, true)) {
            this.queue.add(track);
        }
    }

    public void nextTrack() {
        this.player.startTrack(this.queue.remove(0), false);
    }

    @Override
    public void onTrackEnd(AudioPlayer player, AudioTrack track, AudioTrackEndReason endReason) {
        if (endReason.mayStartNext) {
            nextTrack();
        }
    }

    @Override
    public void onTrackStart(AudioPlayer player, AudioTrack track) {
        if (channel.getGuild().getSelfMember().getPermissions(channel).contains(Permission.MESSAGE_EMBED_LINKS)) {
            MessageEmbed embed = new EmbedBuilder()
                    .setTitle("Now Playing")
                    .addField("Track Name", "[" + track.getInfo().title + "](" + track.getInfo().uri + ")", true)
                    .addField("Length", String.format("%02d min %02d sec",
                            track.getDuration() / TimeUnit.MINUTES.toMillis(1),
                            track.getDuration() % TimeUnit.MINUTES.toMillis(1) / TimeUnit.SECONDS.toMillis(1)
                    ), true)
                    .setThumbnail("https://img.youtube.com/vi/" + track.getIdentifier() + "/default.jpg")
                    .build();
            this.channel.sendMessageEmbeds(embed).queue();
        } else {
            this.channel.sendMessage(track.getInfo().title + " has started playing").queue();
        }
    }

}

File: src/main/kotlin/com/therohankumar/modules/TrackScheduler.kt
package com.therohankumar.modules

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue

class TrackScheduler(private val audioPlayer: AudioPlayer): AudioEventAdapter() {
    val queue: BlockingQueue<AudioTrack> = LinkedBlockingQueue()
    var textChannel: TextChannel? = null
    fun queue(track: AudioTrack) {
        if(audioPlayer.playingTrack == null) {
            audioPlayer.startTrack(track, false)
        }
        else {
            queue.offer(track)
        }
    }

    fun shuffleTrack(){
        val shuffle = queue.shuffled()
        queue.clear()
        queue.addAll(shuffle)
    }

    fun shiftTrack(currentPosittion: Int, newPosition: Int) {
        val queueList = queue.toList()
        val songToMove = queueList[currentPosittion]
        val newQueueList = ArrayList(queueList).apply {
            remove(songToMove)
            val boundedPos = newPosition.coerceIn(0, size)
            add(boundedPos, songToMove)
        }

        queue.clear()
        queue.addAll(newQueueList)
    }

    fun nextTrack() {
        if(queue.isNotEmpty()) audioPlayer.startTrack(queue.poll(), false)
    }

    override fun onTrackEnd(player: AudioPlayer, track: AudioTrack, endReason: AudioTrackEndReason) {
        if (endReason.mayStartNext) {
            nextTrack()
        }
    }

    override fun onTrackStart(player: AudioPlayer, track: AudioTrack) {
        val embed = EmbedUtils.createNowPlayingEmbed(
            trackTitle = track.info.title,
            trackUrl = track.info.uri,
            author = track.info.author,
            durationMillis = track.duration,
            thumbnail = track.info.artworkUrl,
            requestedBy = track.userData as User
        )
        textChannel?.sendMessageEmbeds(embed)?.queue()
    }

    override fun onTrackException(player: AudioPlayer?, track: AudioTrack?, exception: FriendlyException) {
        val embed = EmbedUtils.createErrorEmbed("Track Exception", exception.localizedMessage)
        textChannel?.sendMessageEmbeds(embed)?.queue()
    }
}
